### ISSUE TITLE
NetworkInformationException: Permission denied (.NET 8 + Docker +GCloud)

### ISSUE THREAD

**Issue by christosk92** [dotnet/runtime#94794]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Identical code (below) that works fine in .NET 7, fails with the following error in .NET 8:

```
Unhandled exception. System.Net.NetworkInformation.NetworkInformationException (13): Permission denied
2023-11-16 02:04:53.524 JST
   at System.Net.NetworkInformation.NetworkChange.CreateSocket()
2023-11-16 02:04:53.524 JST
   at System.Net.NetworkInformation.NetworkChange.add_NetworkAddressChanged(NetworkAddressChangedEventHandler value)
2023-11-16 02:04:53.524 JST
   at System.Net.Http.HttpConnectionPoolManager.StartMonitoringNetworkChanges()
2023-11-16 02:04:53.524 JST
   at System.Net.Http.HttpConnectionPool.HandleAltSvc(IEnumerable`1 altSvcHeaderValues, Nullable`1 responseAge)
2023-11-16 02:04:53.524 JST
   at System.Net.Http.HttpConnectionPool.ProcessAltSvc(HttpResponseMessage response)
2023-11-16 02:04:53.524 JST
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
2023-11-16 02:04:53.524 JST
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
2023-11-16 02:04:53.524 JST
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
```

Trimmed and adjusted code (same result for ANY https call):
```cs
var httpClient = new HttpClient();
_ = await httpClient.GetAsync("https://www.google.com");
```

The peculiarity of this issue lies in the fact that the code executes fine on a local environment. But the moment the container is spun up in a google cloud run instance, it fails. 

Another weird thing is that raw TCP connections work fine, we have a database connection that goes directly over TCP, rather than traditional HTTP(s). 

I have spent hours debugging this issue, and sadly we have no other choice but to revert back to .NET 7 for the time being. 

### Expected Behavior

A Proper HTTPs connection

### Exceptions (if any)

NetworkInformationException 

### .NET Version

.NET 8

**Comment by gfoidl** [comment]
> the moment the container is spun up

Can you share which image the container is based?
Is it a non-root image?
Can you try a container that runs as root?

**Comment by wfurt** [comment]
That seems to come from HTTP/3 processing. What do you use as request version @christosk92? 
I think the default is enabled in in .NET 8 but there should be ways how to suppress that. 

https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-http3


Now, if my theory is right, you should see same behavior with .NET 7 _if_ you enable HTTP/3. 
The `NetworkChange.CreateSocket` will attempt to create `Nertlink` socket and it seems like that is not permitted in your container. (you can create simple app to verify that) 
Perhaps something HttpClient can handle better. I feel we should _not_ fail if there is path forward with HTTP/1.1 or H/2.

**Comment by ManickaP** [comment]
This is a general problem for Alt-Svc handling, the spec says you need to clear them on network change [RFC 7838](https://datatracker.ietf.org/doc/html/rfc7838#section-3.1):
>  By default, cached alternative services will be cleared when the
   client detects a network change.  Alternative services that are
   intended to be longer lived (such as those that are not specific to
   the client access network) can carry the "persist" parameter with a
   value "1" as a hint that the service is potentially useful beyond a
   network configuration change.

We just happen to use Alt-Svc only for H/3. So disabling it is a workaround.

I agree with Tomas, we should fix this, at least catch the exception and continue business as usual. If we're not able to detect network changes due to access rights limitations we won't.

**Comment by rzikm** [comment]
As a workaround, setting `DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_HTTP3SUPPORT` env var to `false` should work.

**Comment by ManickaP** [comment]
Or programmatically with [`AppContext`](https://learn.microsoft.com/en-us/dotnet/api/system.appcontext.setswitch?view=net-8.0) switch: "System.Net.SocketsHttpHandler.Http3Support".

**Comment by christosk92** [comment]
Thanks to everyone who took the time to respond to this. We have confirmed that indeed falling back to previous HTTP versions resolves this issue and we’re happily using .NET 8 again !

**Comment by antonfirsov** [comment]
Tentatively triaged to `8.0.x`, we can change the milestone if we decide not to backport it.

**Comment by rzikm** [comment]
opening for 8.0 backport

**Comment by glb-cblin** [comment]
If I undesrstanf correctly, this bug is going to be fixed in February release. Is that correct ?

**Comment by rzikm** [comment]
> If I undesrstand correctly, this bug is going to be fixed in February release. Is that correct ?

Yes, that is the current plan. Meanwhile, you need to use one of the workarounds mentioned above.

**Comment by glb-cblin** [comment]
thanks for the super fast answer @rzikm 👍 

I've used env DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_HTTP3SUPPORT  = "false" and it works

What does it mean for networking/security (with February release) :
1. a non-root container will be able to use HTTP/3 
2. a non-root container will only be able to use HTTP/2 because it needs to be root for HTTP/3

**Comment by rzikm** [comment]
> What does it mean for networking/security (with February release)

option 1. RFC does not require clients to do network monitoring. So in case we cannot detect network changes (due to constrained environment), the alt-svc headers will still get invalidated by the specified timeout.

**Comment by christosk92** [comment]
> > What does it mean for networking/security (with February release)
> 
> option 1. RFC does not require clients to do network monitoring. So in case we cannot detect network changes (due to constrained environment), the alt-svc headers will still get invalidated by the timeout specified in the header.
> 
> Note also that for HTTP/3 to work, you need to install MsQuic on the machine (on Linux, on windows the dll is distributed with the runtime).

So this means dockerfiles should be adjusted accordingly to install MsQuic? Or are they already included if youre using the msft docker images to build your .net container?

**Comment by rzikm** [comment]
> So this means dockerfiles should be adjusted accordingly to install MsQuic? Or are they already included if youre using the msft docker images to build your .net container?

Yes, published docker images do not install msquic and it needs to be installed separately. MsQuic does not publish official packages for all distributions supported by .NET and thus installing it nowhere by default is better than installing it only on some. Also, not all applications make use of QUIC or HTTP/3 so it makes sense for user to specifically opt in for this feature.

**Comment by christosk92** [comment]
Makes sense, thanks for the clarification.

**Comment by rzikm** [comment]
Will be fixed in 8.0.2 release: https://github.com/dotnet/runtime/pull/95123

**Comment by nthonymiller** [comment]
> As a workaround, setting `DOTNET_SYSTEM_NET_HTTP_SOCKETSHTTPHANDLER_HTTP3SUPPORT` env var to `false` should work.

Thanks for workaround, I ran into the exact same issue except I was using FirebaseAdmin package to validate an JWT.

```
  decodedToken = await firebaseAuth.VerifyIdTokenAsync(idToken);
```

**Comment by karelz** [comment]
Fixed in main (9.0) in PR #94861 and in 8.0.2 in PR #95123.
